core/src/OpenACC/Kokkos_OpenACCSpace.cpp

```diff
@@ -0,0 +1,204 @@
+//
+// 3. Neither the name of the Corporation nor the names of the
+// contributors may be used to endorse or promote products derived from
+// this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY NTESS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+// PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL NTESS OR THE
+// CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+// EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+// PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+// LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+// NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+// SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#ifndef KOKKOS_IMPL_PUBLIC_INCLUDE
+#define KOKKOS_IMPL_PUBLIC_INCLUDE
+#endif
+
+#include <Kokkos_Macros.hpp>
+#if defined KOKKOS_ENABLE_OPENACC
+
+#include <openacc.h>
+
+/*--------------------------------------------------------------------------*/
+
+#include <Kokkos_OpenACC.hpp>
+#include <OpenACC/Kokkos_OpenACC_Exec.hpp>
+#include <Kokkos_OpenACCSpace.hpp>
+#include <impl/Kokkos_MemorySpace.hpp>
+
+//----------------------------------------------------------------------------
+//----------------------------------------------------------------------------
+
+namespace Kokkos {
+namespace Experimental {
+/* Default allocation mechanism */
+OpenACCSpace::OpenACCSpace() {}
+
+void *OpenACCSpace::allocate(const size_t arg_alloc_size) const {
+  return allocate("[unlabeled]", arg_alloc_size);
+}
+void *OpenACCSpace::allocate(const char *arg_label, const size_t arg_alloc_size,
+                             const size_t arg_logical_size) const {
+  return impl_allocate(arg_label, arg_alloc_size, arg_logical_size);
+}
+void *OpenACCSpace::impl_allocate(
+    const char *arg_label, const size_t arg_alloc_size,
+    const size_t arg_logical_size,
+    const Kokkos::Tools::SpaceHandle arg_handle) const {
+  static_assert(sizeof(void *) == sizeof(uintptr_t),
+                "Error sizeof(void*) != sizeof(uintptr_t)");
+
+  void *ptr = nullptr;
+
+  //[DEBUG] Disabled due to the synchronous behavior of the current
+  // implementation.
+  /*
+    OpenACC::impl_static_fence(
+        "Kokkos::OpenACCSpace::impl_allocate: Pre OpenACC Allocation");
+  */
+
+  ptr = acc_malloc(arg_alloc_size);
+
+  //[DEBUG] Disabled due to the synchronous behavior of the current
+  // implementation.
+  /*
+    OpenACC::impl_static_fence(
+        "Kokkos::OpenACCSpace::impl_allocate: Post OpenACC Allocation");
+  */
```

Comment:
What are you planning to do with this before merging? We would not normally use `FIXEM_OPENACC` for stuff to be fixed later to have something uniform to search for.

Do you need the fences anyway? Isn't `acc_malloc` synchronous?

Reply:
`acc_malloc` is synchronous, and thus we don't need fences. I added these comments as placeholders, since I don't fully understand the exact synchronization points (and scopes) in the Kokkos execution model with multiple device instance support.

Reply:
Those placeholder comments are removed from this PR.